ddl/syncer.go

```diff
@@ -153,6 +169,92 @@ func (s *schemaVersionSyncer) Init(ctx context.Context) error {
 	return errors.Trace(err)
 }
 
+// GetServerInfoFromPD implements SchemaSyncer.GetServerInfoFromPD interface.
+func (s *schemaVersionSyncer) GetServerInfoFromPD(ctx context.Context, id string) (*util.DDLServerInfo, error) {
+	var err error
+	var resp *clientv3.GetResponse
+	ddlPath := fmt.Sprintf("%s/%s", DDLServerInformation, id)
+	for {
+		if isContextDone(ctx) {
+			err = errors.Trace(ctx.Err())
+			return nil, err
+		}
+
+		resp, err = s.etcdCli.Get(ctx, ddlPath)
+		if err != nil {
+			log.Infof("[syncer] get ddl server info,  ddl %s failed %v, continue checking.", ddlPath, err)
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+		if len(resp.Kvs) > 0 {
+			info := &util.DDLServerInfo{}
+			err := json.Unmarshal(resp.Kvs[0].Value, info)
+			if err != nil {
+				log.Infof("[syncer] get ddl server info, ddl %s json.Unmarshal %v failed %v.", resp.Kvs[0].Key, resp.Kvs[0].Value, err)
+				return nil, errors.Trace(err)
+			}
+			return info, nil
+		}
+	}
+}
+
+// GetAllServerInfoFromPD implements SchemaSyncer.GetAllServerInfoFromPD interface.
+func (s *schemaVersionSyncer) GetAllServerInfoFromPD(ctx context.Context) (map[string]*util.DDLServerInfo, error) {
+	var err error
+	allDDLInfo := make(map[string]*util.DDLServerInfo)
+	for {
+		if isContextDone(ctx) {
+			// ctx is canceled or timeout.
+			err = errors.Trace(ctx.Err())
+			return nil, err
+		}
+
+		resp, err := s.etcdCli.Get(ctx, DDLServerInformation, clientv3.WithPrefix())
+		if err != nil {
+			log.Infof("[syncer] get all ddl server info failed %v, continue checking.", err)
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+
+		for _, kv := range resp.Kvs {
+			info := &util.DDLServerInfo{}
+			err := json.Unmarshal(kv.Value, info)
+			if err != nil {
+				log.Infof("[syncer] get all ddl server info, ddl %s json.Unmarshal %v failed %v.", kv.Key, kv.Value, err)
+				return nil, errors.Trace(err)
+			}
+			allDDLInfo[info.ID] = info
+		}
+		return allDDLInfo, nil
+	}
+}
+
+// StoreSelfServerInfoToPD implements SchemaSyncer.StoreSelfServerInfoToPD interface.
+func (s *schemaVersionSyncer) StoreSelfServerInfoToPD(ctx context.Context, info *util.DDLServerInfo) error {
+	infoBuf, err := json.Marshal(info)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = s.putKV(ctx, keyOpDefaultRetryCnt, s.selfServerInfoPath, hack.String(infoBuf))
+	return errors.Trace(err)
+}
+
+// RemoveSelfServerInfoFromPD implements SchemaSyncer.RemoveSelfServerInfoFromPD interface.
+func (s *schemaVersionSyncer) RemoveSelfServerInfoFromPD() error {
+	var err error
+	ctx := context.Background()
+	for i := 0; i < keyOpDefaultRetryCnt; i++ {
+		childCtx, cancel := context.WithTimeout(ctx, keyOpDefaultTimeout)
+		_, err = s.etcdCli.Delete(childCtx, s.selfServerInfoPath)
+		cancel()
+		if err == nil {
+			return nil
+		}
+		log.Warnf("[syncer] remove server info path %s failed %v no.%d", s.selfServerInfoPath, err, i)
+	}
```

Comment:
Cloud this(line176-198) be extracted into a function？ It will be used in RemoveSelfVersionPath too.